lib/fluent/plugin/out_google_cloud.rb

```diff
@@ -873,15 +641,409 @@ def fetch_gce_metadata(metadata_path)
+    def set_project_id
+      @project_id ||= fetch_gce_metadata('project/project-id') if
+        @platform == Platform::GCE
+      @project_id ||= CredentialsInfo.project_id
+    rescue StandardError => e
+      @log.error 'Failed to obtain project id: ', error: e
+    end
+
+    # 1. Return the value if it is explicitly set in the config already.
+    # 2. If not, try to retrieve it by calling metadata servers directly.
+    def set_vm_id
+      @vm_id ||= fetch_gce_metadata('instance/id') if @platform == Platform::GCE
+      @vm_id ||= ec2_metadata['instanceId'] if @platform == Platform::EC2
+    rescue StandardError => e
+      @log.error 'Failed to obtain vm_id: ', error: e
+    end
+
+    # 1. Return the value if it is explicitly set in the config already.
+    # 2. If not, try to retrieve it locally.
+    def set_vm_name
+      @vm_name ||= Socket.gethostname
+    rescue StandardError => e
+      @log.error 'Failed to obtain vm name: ', error: e
+    end
+
+    # 1. Return the value if it is explicitly set in the config already.
+    # 2. If not, try to retrieve it locally.
+    def set_location
+      # Response format: "projects/<number>/zones/<zone>"
+      @zone ||= fetch_gce_metadata('instance/zone').rpartition('/')[2] if
+        @platform == Platform::GCE
+      @zone ||= 'aws:' + ec2_metadata['availabilityZone'] if
+        @platform == Platform::EC2 && ec2_metadata.key?('availabilityZone')
+    rescue StandardError => e
+      @log.error 'Failed to obtain location: ', error: e
+    end
+
+    # Retrieve monitored resource via the legacy way.
+    #
+    # TODO(qingling128): Use this as only a fallback plan after Metadata Agent
+    # support is added.
+    def determine_agent_level_monitored_resource_via_legacy
+      resource = Google::Apis::LoggingV2beta1::MonitoredResource.new(
+        labels: {})
+      resource.type = determine_agent_level_monitored_resource_type
+      resource.labels = determine_agent_level_monitored_resource_labels(
+        resource.type)
+      resource
+    end
+
+    # Determine agent level monitored resource type.
+    def determine_agent_level_monitored_resource_type
+      case @platform
+      when Platform::OTHER
+        # Unknown platform will be defaulted to GCE instance.
+        return COMPUTE_CONSTANTS[:resource_type]
+
+      when Platform::EC2
+        return EC2_CONSTANTS[:resource_type]
+
+      when Platform::GCE
+        # Resource types determined by @subservice_name config.
+        if @subservice_name
+          SUB_SERVICE_LIST.each do |resource_constants|
+            return resource_constants[:resource_type] if
+              @subservice_name == resource_constants[:service]
+          end
+          # Default back to GCE if invalid value is detected.
+          return COMPUTE_CONSTANTS[:resource_type]
+        end
+
+        # Resource types determined by @detect_subservice config.
+        if @detect_subservice
```

Comment:
I renamed the maps to avoid confusion.

Also changed
`return resource_type if expected_attributes.all? { |attr| attributes.include?(attr) }`
to 
`return resource_type if attributes.superset?(expected)`.

I don't get why we should put `service -> attributes` in `SUBSERVICE_METADATA_ATTRIBUTES`, then do another map lookup to get `service -> resource_type` though. Can't we just put `resource_type -> attributes` in `SUBSERVICE_METADATA_ATTRIBUTES`? Ultimately we want to get the resource types.